### PR TITLE
ARROW-10313: [C++] Faster UTF8 validation for small strings

### PR DESCRIPTION
**Pull request by pitrou**
This improves CSV string conversion performance by about 30%.

### CODE REVIEW

**Comment by pitrou** [comment]
This also improves the ARROW-10308 benchmark by about 9%.

**Comment by pitrou** [comment]
Another possibility would be to compute and store ASCIIness of values while parsing CSV (ideally this should not cost anything CPU-wise). Then we can reuse that information to skip UTF8 validation for most values.

File: cpp/src/arrow/util/utf8.h [hunk 1]
```diff
@@ -154,13 +156,52 @@ inline bool ValidateUTF8(const uint8_t* data, int64_t size) {
...
 
-  // Validate string tail one byte at a time
+  // Check if string tail is full ASCII (common case, fast)
+  if (size >= 4) {
+    uint32_t mask1, mask2;
+    memcpy(&mask2, data + size - 4, 4);
+    memcpy(&mask1, data, 4);
+    if (ARROW_PREDICT_TRUE(((mask1 | mask2) & high_bits_32) == 0)) {
+      return true;
+    }
```
Comment (bkietz) [inline cpp/src/arrow/util/utf8.h:166]:
```suggestion
    uint32_t head_mask = internal::SafeLoadAs<uint32_t>(data);
    uint32_t tail_mask = internal::SafeLoadAs<uint32_t>(data + size - 4);
    if (ARROW_PREDICT_TRUE(((head_mask | tail_mask) & high_bits_32) == 0)) {
      return true;
    } 
```

**Review by bkietz — CHANGES_REQUESTED** [review]
LGTM, just one nit

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/util/utf8.h [hunk 2]
```diff
@@ -154,13 +157,50 @@ inline bool ValidateUTF8(const uint8_t* data, int64_t size) {
...
+  // Check if string tail is full ASCII (common case, fast)
+  if (size >= 4) {
+    uint32_t tail_mask = SafeLoadAs<uint32_t>(data + size - 4);
+    uint32_t head_mask = SafeLoadAs<uint32_t>(data);
+    if (ARROW_PREDICT_TRUE(((head_mask | tail_mask) & high_bits_32) == 0)) {
+      return true;
+    }
+  } else if (size >= 2) {
+    uint16_t tail_mask = SafeLoadAs<uint32_t>(data + size - 2);
+    uint16_t head_mask = SafeLoadAs<uint32_t>(data);
```
Comment (bkietz) [inline cpp/src/arrow/util/utf8.h:169]:
```suggestion
    uint16_t tail_mask = SafeLoadAs<uint16_t>(data + size - 2);
    uint16_t head_mask = SafeLoadAs<uint16_t>(data);
```

File: cpp/src/arrow/util/utf8.h [hunk 3]
```diff
@@ -98,8 +100,9 @@ inline bool ValidateUTF8(const uint8_t* data, int64_t size) {
     // XXX This is doing an unaligned access.  Contemporary architectures
     // (x86-64, AArch64, PPC64) support it natively and often have good
     // performance nevertheless.
-    memcpy(&mask, data, 8);
-    if (ARROW_PREDICT_TRUE((mask & high_bits_64) == 0)) {
+    uint64_t mask64;
+    memcpy(&mask64, data, 8);
```
Comment (bkietz) [inline cpp/src/arrow/util/utf8.h:104]:
```suggestion
    uint64_t mask64 = SafeLoadAs<uint64_t>(data);
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Whoops

**Pull request closed without merging**